<a href="https://colab.research.google.com/github/juanpalominoh/Fundamentos-de-Econometria/blob/main/Tema_3_Modelo_en_Desviaciones.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tema 3: Modelo en Desviaciones Respecto a la Media

In [ ]:
install.packages("readxl")
library("readxl")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
base <- as.data.frame(read_excel("/content/Ejemplo1.xlsx"))
base

Id X  Y   
1   1  6  230
2   2  7  240
3   3 13  250
4   4 10  400
5   5 16  670
6   6 12  335
7   7 14  400
8   8 13  550
9   9 13  850
10 10 12  234
11 11 18 1000
12 12 15  345
13 13 14  890
14 14 14  450
15 15 12  425

Calcular las desviaciones con respecto a la media de $X$ e $Y$

In [ ]:
# 1. Se calcula los promedios
mX <- mean(base$X)
mY <- mean(base$Y)
mX
mY

[1] 12.6

[1] 484.6

In [ ]:
# 2. Se calcula las desviaciones
dX <- base$X - mX
dY <- base$Y - mY

Calcular $xy$ e $x^{2}$

In [ ]:
xy <- dX*dY

In [ ]:
sq_X <- dX^2

In [ ]:
# Agregando a la base
base_2 <- cbind(base, mX, dX, mY, dY, sq_X, xy)
base_2

Id X  Y    mX   dX   mY    dY     sq_X  xy     
1   1  6  230 12.6 -6.6 484.6 -254.6 43.56 1680.36
2   2  7  240 12.6 -5.6 484.6 -244.6 31.36 1369.76
3   3 13  250 12.6  0.4 484.6 -234.6  0.16  -93.84
4   4 10  400 12.6 -2.6 484.6  -84.6  6.76  219.96
5   5 16  670 12.6  3.4 484.6  185.4 11.56  630.36
6   6 12  335 12.6 -0.6 484.6 -149.6  0.36   89.76
7   7 14  400 12.6  1.4 484.6  -84.6  1.96 -118.44
8   8 13  550 12.6  0.4 484.6   65.4  0.16   26.16
9   9 13  850 12.6  0.4 484.6  365.4  0.16  146.16
10 10 12  234 12.6 -0.6 484.6 -250.6  0.36  150.36
11 11 18 1000 12.6  5.4 484.6  515.4 29.16 2783.16
12 12 15  345 12.6  2.4 484.6 -139.6  5.76 -335.04
13 13 14  890 12.6  1.4 484.6  405.4  1.96  567.56
14 14 14  450 12.6  1.4 484.6  -34.6  1.96  -48.44
15 15 12  425 12.6 -0.6 484.6  -59.6  0.36   35.76

In [ ]:
# Por propiedad, la suma de las desviaciones debe ser igual a cero.
options(scipen=999)
sum(dX)
sum(dY)

[1] 0.000000000000005329071

[1] -0.0000000000003410605

Calcular el coefiente $\hat{\beta}_{1}$ y $\hat{\beta}_{2}$

In [ ]:
sum(xy)
sum(sq_X)

[1] 7103.6

[1] 135.6

In [ ]:
beta2 <- sum(xy)/sum(sq_X )

In [ ]:
beta1 <- mY - (beta2*mX)

In [ ]:
beta1
beta2

[1] -175.469

[1] 52.38643

Calcular $\hat{Y}_{i}$, los residuos del modelo, y los residuos al cuadrado.

\begin{gather*}
\hat{Y}_{i}=-175.469 + 52.3864X_{i}
\end{gather*}

In [ ]:
# Calcular el predicho de Y
Y_hat <- beta1 + beta2*base_2$X

In [ ]:
# Calcular los residuos
e <- base_2$Y - Y_hat

In [ ]:
# Calcular los residuos al cuadrado
sq_e <- e^2

In [ ]:
cbind(base_2, Y_hat, e, sq_e)

Id X  Y    mX   dX   mY    dY     sq_X  xy      Y_hat    e          
1   1  6  230 12.6 -6.6 484.6 -254.6 43.56 1680.36 138.8496   91.150442
2   2  7  240 12.6 -5.6 484.6 -244.6 31.36 1369.76 191.2360   48.764012
3   3 13  250 12.6  0.4 484.6 -234.6  0.16  -93.84 505.5546 -255.554572
4   4 10  400 12.6 -2.6 484.6  -84.6  6.76  219.96 348.3953   51.604720
5   5 16  670 12.6  3.4 484.6  185.4 11.56  630.36 662.7139    7.286136
6   6 12  335 12.6 -0.6 484.6 -149.6  0.36   89.76 453.1681 -118.168142
7   7 14  400 12.6  1.4 484.6  -84.6  1.96 -118.44 557.9410 -157.941003
8   8 13  550 12.6  0.4 484.6   65.4  0.16   26.16 505.5546   44.445428
9   9 13  850 12.6  0.4 484.6  365.4  0.16  146.16 505.5546  344.445428
10 10 12  234 12.6 -0.6 484.6 -250.6  0.36  150.36 453.1681 -219.168142
11 11 18 1000 12.6  5.4 484.6  515.4 29.16 2783.16 767.4867  232.513274
12 12 15  345 12.6  2.4 484.6 -139.6  5.76 -335.04 610.3274 -265.327434
13 13 14  890 12.6  1.4 484.6  405.4  1.96  567.56 557.9410  332.058997
14 14 14  450 12.6  1.4 484.6  -34.6  1.96  -48.44 557.9410 -107.941003
15 15 12  425 12.6 -0.6 484.6  -59.6  0.36   35.76 453.1681  -28.168142
   sq_e        
1    8308.40316
2    2377.92885
3   65308.13941
4    2663.04710
5      53.08777
6   13963.70969
7   24945.36041
8    1975.39605
9  118642.65268
10  48034.67429
11  54062.42274
12  70398.64704
13 110263.17752
14  11651.26012
15    793.44420

In [ ]:
# La suma de los residuos es cero
sum(e)

[1] -0.0000000000001705303

In [ ]:
# La Sumatoria de residuos al cuadrado (SCR)
sum(sq_e)

[1] 533441.4

Calcular el estimador de $\sigma^{2}$:

\begin{gather*}
\hat{\sigma^{2}}= \frac{\sum \hat{\epsilon}_{i}^{2}}{n-2}
\end{gather*}

In [ ]:
sigma2_hat <- sum(sq_e)/(nrow(base_2)-2)
sigma2_hat

[1] 41033.95

Calcular las varianzas de $\hat{\beta}_{1}$ y $\hat{\beta}_{2}$. 

\begin{gather*}
\hat{Var}(\hat{\beta}_{1})=\sigma^{2}(\frac{1}{n}+\frac{\bar{X}^{2}}{\sum x^{2}_{i}})
\end{gather*}

\begin{gather*}
\hat{Var}(\hat{\beta}_{2})=\frac{\sigma^{2}}{\sum x_{i}^{2}}
\end{gather*}

In [ ]:
# Calcular varianza de beta1
var_beta1 <- sigma2_hat*((1/nrow(base_2)) + ((mX^2)/sum(sq_X)))
var_beta1

[1] 50778

In [ ]:
# Calcular varianza de beta2
var_beta2 <- sigma2_hat / sum(sq_X)
var_beta2

[1] 302.6103

Estimar una regresión de $Y$ sobre $X$

In [ ]:
model <- lm(Y ~ X, data=base_2)
summary(model)


Call:
lm(formula = Y ~ X, data = base_2)

Residuals:
    Min      1Q  Median      3Q     Max 
-265.33 -138.05    7.29   71.38  344.45 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept)  -175.47     225.34  -0.779     0.45  
X              52.39      17.40   3.011     0.01 *
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 202.6 on 13 degrees of freedom
Multiple R-squared:  0.4109,	Adjusted R-squared:  0.3656 
F-statistic: 9.069 on 1 and 13 DF,  p-value: 0.01002


In [ ]:
225.34^2
17.4^2

[1] 50778.12

[1] 302.76

Hallar SCT, SCE y SCR

In [ ]:
# SCR
scr <- sum((fitted(model) - base_2$Y)^2)
scr

[1] 533441.4

In [ ]:
# SCE
sce <- sum((fitted(model) - mean(base_2$Y))^2)
sce

[1] 372132.2

In [ ]:
# SCT
sct <- scr + sce
sct

[1] 905573.6

Calcular $R^{2}$

In [ ]:
# R_squared
sce / sct

[1] 0.4109354